### PR TITLE
Transaction link: copy link, text and more

### LINKED ISSUE #1885: 🚀 [Feature] Copy the Link Text as well
## 🚀 Feature
Have another button to copy both, the link and the text. This way, when pasting the link in an email the user does not have to repeat the text.

Fields to copy:
 * Amount 
 * Sender
 *  Text 
 *  Link 
 *  Expired Date

### PR DESCRIPTION
**Pull request by mahula**
## 🍰 Pullrequest
In the transaction overview a new dropdown item is introcuded for copying a transaction link.
When uses, besides the link also the following are copied

- Amount
- Sender
- Sender's memo
- Expired Date

The copied data is written to the clipboard in a nice human readable format in the locale use by the user:

```
<link>
<sender's firstname> sends you <amount> Gradido.
"<sender's memo text>"
To credit your Gradido, click on the link! It is valid <validation date>.
```

### Issues

- fixes #1885

### CODE REVIEW

File: frontend/src/locales/de.json [hunk 1]
```diff
@@ -125,7 +125,9 @@
   "gdd_per_link": {
     "choose-amount": "Wähle einen Betrag aus, welchen du per Link versenden möchtest. Du kannst auch noch eine Nachricht eintragen. Beim Klick „Jetzt generieren“ wird ein Link erstellt, den du versenden kannst.",
     "copy": "kopieren",
+    "copy-with-text": "Link und Text kopieren",
     "created": "Der Link wurde erstellt!",
+    "credit-your-gradido": "Um deine Gradido gutzuschreiben, klicke auf den Link!",
```
Comment (ogerly) [inline frontend/src/locales/de.json:130]:
```suggestion
    "credit-your-gradido": "Damit die GDD gutgeschrieben werden können, klicke auf den Link!",
```

File: frontend/src/components/TransactionLinks/TransactionLink.vue [hunk 2]
```diff
@@ -99,6 +103,22 @@ export default {
           this.toastError(this.$t('gdd_per_link.not-copied'))
         })
     },
+    copyLinkWithText() {
+      navigator.clipboard
+        .writeText(
+          `${this.link}\n` +
+          `${this.$store.state.firstName} ${this.$t('transaction-link.send_you')} ${this.amount} Gradido.\n` +
+          `\"${this.memo}\"\n` +
+          `${this.$t('gdd_per_link.credit-your-gradido')} ${this.$t('gdd_per_link.validUntilDate', {date: this.$d(new Date(this.validUntil), 'short')})}`
```
Comment (ogerly) [inline frontend/src/components/TransactionLinks/TransactionLink.vue]:
```suggestion
           `${this.link}\n
            ${this.$store.state.firstName} ${this.$t('transaction-link.send_you')} ${
            this.amount
          }. GDD \n
            "${this.memo}"\n
            ${this.$t('gdd_per_link.credit-your-gradido')} ${this.$t(
            'gdd_per_link.validUntilDate',
            { date: this.$d(new Date(this.validUntil), 'short') },
          )}`
```

Comment (ogerly) [reply frontend/src/components/TransactionLinks/TransactionLink.vue]: the pluses are superfluous. a single apostrophe at the beginning and end is sufficient.

File: frontend/src/components/TransactionLinks/TransactionLink.vue [hunk 3]
```diff
@@ -22,6 +22,10 @@
                 <b-icon icon="clipboard"></b-icon>
                 {{ $t('gdd_per_link.copy') }}
               </b-dropdown-item>
+              <b-dropdown-item v-if="validLink" class="test-copy-link" @click="copyLinkWithText()">
```
Comment (ogerly) [inline frontend/src/components/TransactionLinks/TransactionLink.vue:25]:
```suggestion
              <b-dropdown-item v-if="validLink" class="test-copy-text pt-3" @click="copyLinkWithText()">
```

Comment (ogerly) [reply frontend/src/components/TransactionLinks/TransactionLink.vue:25]: this is used in the tests to find the click

Comment (ogerly) [reply frontend/src/components/TransactionLinks/TransactionLink.vue:25]: **pt-3** ist der **padding-top** von bootstrap

das die abstände passen 

image

File: frontend/src/locales/de.json [hunk 4]
```diff
@@ -149,7 +152,8 @@
     "redeemed-title": "eingelöst",
     "to-login": "Log dich ein",
     "to-register": "Registriere ein neues Konto.",
-    "validUntil": "Gültig bis"
+    "validUntil": "Gültig bis",
+    "validUntilDate": "Sie sind bis zum {date} gültig."
```
Comment (ogerly) [inline frontend/src/locales/de.json:156]:
```suggestion
    "validUntilDate": "Der Link ist bis zum {date} gültig."
```

File: frontend/src/locales/en.json [hunk 5]
```diff
@@ -149,7 +152,8 @@
     "redeemed-title": "redeemed",
     "to-login": "Log in",
     "to-register": "Register a new account.",
-    "validUntil": "Valid until"
+    "validUntil": "Valid until",
+    "validUntilDate": "They are valid until {date}."
```
Comment (ogerly) [inline frontend/src/locales/en.json:156]:
```suggestion
    "validUntilDate": "The link is valid until {date}."
```

File: frontend/src/locales/en.json [hunk 6]
```diff
@@ -125,7 +125,9 @@
   "gdd_per_link": {
     "choose-amount": "Select an amount that you would like to send via link. You can also enter a message. Click 'Generate now' to create a link that you can share.",
     "copy": "copy",
+    "copy-with-text": "copy link and text",
     "created": "Link was created!",
+    "credit-your-gradido": "To credit your Gradido, click on the link!",
```
Comment (ogerly) [inline frontend/src/locales/en.json:130]:
```suggestion
    "credit-your-gradido": "For the GDD to be credited, click on the link!",
```

**Review by ogerly — CHANGES_REQUESTED** [review]

File: frontend/src/components/TransactionLinks/TransactionLink.vue [hunk 7]
```diff
@@ -99,6 +103,26 @@ export default {
...
+        .writeText(
+          `${this.link}\n` +
+            `${this.$store.state.firstName} ${this.$t('transaction-link.send_you')} ${
+              this.amount
+              } Gradido.\n` +
+            `"${this.memo}"\n` +
+            `${this.$t('gdd_per_link.credit-your-gradido')} ${this.$t(
+              'gdd_per_link.validUntilDate', {
+                date: this.$d(new Date(this.validUntil), 'short')})
+                }`)
```
Comment (ogerly) [inline frontend/src/components/TransactionLinks/TransactionLink.vue:117]:
```suggestion
          ``${this.link}\n
            ${this.$store.state.firstName} ${this.$t('transaction-link.send_you')} ${
            this.amount
          }. Gradido \n
            "${this.memo}"\n
            ${this.$t('gdd_per_link.credit-your-gradido')} ${this.$t(
            'gdd_per_link.validUntilDate',
            { date: this.$d(new Date(this.validUntil), 'short') },
          )}`)
```

**Review by ogerly — CHANGES_REQUESTED** [review]

File: frontend/src/components/TransactionLinks/TransactionLink.vue (hunk 2 above)
Comment (mahula) [reply frontend/src/components/TransactionLinks/TransactionLink.vue]: We solved this in another suggestion.

File: frontend/src/locales/en.json [hunk 8]
```diff
@@ -125,14 +125,17 @@
   "gdd_per_link": {
     "choose-amount": "Select an amount that you would like to send via link. You can also enter a message. Click 'Generate now' to create a link that you can share.",
     "copy": "copy",
+    "copy-with-text": "copy link and text",
```
Comment (Mogge) [inline frontend/src/locales/en.json:128]:
```suggestion
    "copy-with-text": "Copy link and text",
```

File: frontend/src/components/TransactionLinks/TransactionLink.vue [hunk 9]
```diff
@@ -99,6 +107,27 @@ export default {
...
+          `${this.link}\n` +
+          `${this.$store.state.firstName} ${this.$t('transaction-link.send_you')} ${
+            this.amount
+          } Gradido.\n` +
+          `"${this.memo}"\n`+
+          `${this.$t('gdd_per_link.credit-your-gradido')} ${this.$t(
+            'gdd_per_link.validUntilDate',
+            { date: this.$d(new Date(this.validUntil), 'short') },
+          )}`,
+        )
```
Comment (Mogge) [inline frontend/src/components/TransactionLinks/TransactionLink.vue:122]:
```suggestion
        .writeText(
          `${this.link}\n
${this.$store.state.firstName} ${this.$t('transaction-link.send_you')} ${this.amount} Gradido.
"${this.memo}"
${this.$t('gdd_per_link.credit-your-gradido')} ${this.$t('gdd_per_link.validUntilDate',  { date: this.$d(new Date(this.validUntil), 'short') })}`,
        )
```

**Review by Mogge — CHANGES_REQUESTED** [review]

File: frontend/src/components/TransactionLinks/TransactionLink.spec.js [hunk 10]
```diff
@@ -9,22 +9,24 @@ const mockAPIcall = jest.fn()
...
+    },
+  },
 }
 
 const propsData = {
   amount: '75',
   link: 'http://localhost/redeem/c00000000c000000c0000',
   holdAvailableAmount: '5.13109484759482747111',
   id: 12,
+  firstname: 'Testy',
```
Comment (Mogge) [inline frontend/src/components/TransactionLinks/TransactionLink.spec.js:29]: Not needed?

**Review by Mogge — CHANGES_REQUESTED** [review]

**Review by ogerly — DISMISSED** [review]

File: frontend/src/components/TransactionLinks/TransactionLink.spec.js (hunk 10 above)
Comment (mahula) [reply frontend/src/components/TransactionLinks/TransactionLink.spec.js:29]: True

Comment (Mogge) [inline frontend/src/components/TransactionLinks/TransactionLink.spec.js:29]: So why it is still there?

**Review by Mogge — DISMISSED** [review]

File: frontend/src/components/TransactionLinks/TransactionLink.spec.js (hunk 10 above)
Comment (Mogge) [reply frontend/src/components/TransactionLinks/TransactionLink.spec.js:29]: @mahula But it is still there. Please commit without this line again.

**Review by Mogge — APPROVED** [review]

**Pull request merged**